### ISSUE TITLE
Allow dependency name overriding or elimination

### ISSUE THREAD

**Issue by leorochael** [astral-sh/uv#4422]
While `uv` now has overrides, and issue #2686 is suggesting being able to declare certain overrides as pertaining to a specific dependency, I have an additional suggestion/request.

I would like to be able to override a package dependency with a package that has a different name.

As a concrete case, the [FuelSDK](https://pypi.org/project/FuelSDK/) package [depends](https://github.com/salesforce-marketingcloud/FuelSDK-Python/blob/master/setup.py#L20) on `suds-jurko`, but has been tested to work with it's fork/successor project [`suds-community`](https://github.com/suds-community/suds), a.k.a. just [`suds`](https://pypi.org/project/suds/) these days.

(Yes, the project has come full circle, with the fork of the fork officially taking over the original name).

A possibility that comes to mind would be a special syntax for the `-o overrides.txt` file like:

```plain
suds-jurko=>suds>=1.1,<2.0
```

Or perhaps:

```plains
suds-jurko:suds>=1.1,<2.0
```

Considering the long history of python packages that fork and continue the work of orphaned packages (e.g. PIL/Pillow), this could be a nice complement to the overrides functionality of uv.

Perhaps it could even be used to remove a dependency completely, e.g.:

* `suds-jurko=>!`
* `suds-jurko=>`
* `suds-jurko:!`
* `suds-jurko:`

**Comment by leorochael** [comment]
Thinking about this some more, just the functionality of blocking a dependency from being installed already offers a way to do this kind of override.

That is, I could add to my `requirements.in`:

```
Fuel-SDK
suds>=1.1,<2.0
```

Then to my `overrides` one of the options for deleting a removing a dependency I mentioned previously, or perhaps just a special version. e.g one of:

* `suds-jurko==skip`
* `suds-jurko==none`
* `suds-jurko==`

**Comment by zanieb** [comment]
Interesting idea! Makes some sense, though I'm also not quite sure how it should be expressed and if it's feasible.

**Comment by leorochael** [comment]
Another option to express dependency elision is to use the URI version specifier to mean: install this package from nowhere.

* `suds-jurko@skip:`
* `suds-jurko@about:blank`

The last one was suggested to me by github copilot:

> To explicitly reference nothing or create a URI that signifies an absence of a resource, you can use the `about:blank` URI. This URI leads to an empty document in web browsers and is often used to represent a null or non-existent resource in contexts where a URI is required but there is no actual resource to point to.

**Comment by leorochael** [comment]
And since URIs effectively allow namespacing on schemes, and the definition of the interpretation of schemes is up to whoever introduces them, we could perhaps use schemes to implement the package replacement, e.g:

* `suds-jurko@replacement:suds>=1.1,<2.0`

**Comment by leorochael** [comment]
Or, if we want to follow examples of other tools which use prefixed schemes, like `git+`, we could namespace the scheme with `uv+`, like:

* `suds-jurko@uv+replacement:suds>=1.1,<2.0`

**Comment by leorochael** [comment]
Note that `--no-emit-package/--unsafe-package suds-jurko` doesn't help here, because `suds-jurko` is broken at package building, so resolution still fails.

**Comment by leorochael** [comment]
An alternative that would help me would be to just ignore the sub-dependencies of a package on a per-package basis.

That is, instead of having an elision or renaming on my overrides file, I would like to add something like this in my `requirements.in`:

```
# FuelSDK has broken sub-dependencies:
--no-deps-for FuelSDK
FuelSDK
# Corrected FuelSDK dependencies manually added here:
pyjwt >= 1.5.3
requests >= 2.18.4
suds >= 1.1.2
```

For reference, such an option has been under discussion for a while in pypa/pip#9948

**Comment by charliermarsh** [comment]
This seems like a reasonable use-case to me.

**Comment by hauntsaninja** [comment]
I think this would be very useful! At work, we have systems like this for first party packages, but not third party packages. It would make some things less painful if this were the case and unblock some other future use cases.

A related feature request would be the ability to override edges, not just nodes. E.g., if I know one package is overpinning a dependency in a way that I know is safe to override, I might still want to resolve older or error if something else pins it.

I don't have strong opinions on syntax. => is a good suggestion, but could be typo-ed. Maybe using `@@` as a separator? lhs is the thing to be overriden, rhs is empty or PEP 508
```
node>=2
node @@ node>=2  # same as above
node @@ node_nextgen>=2  # replace node with node_nextgen
node @@  # eliminate node (and don't resolve its deps)
node -> dep @@ dep>=2  # override node's dependency on dep
```
If syntax is controversial, another option is to switch to TOML or something, especially if uv has plans to use such a format somewhere else.

**Comment by charliermarsh** [comment]
I realized that this actually is possible today... You can use a never-truthy marker. For example, to remove `typing-extensions`, use an `overrides.txt` like:

```
typing-extensions ; sys_platform == 'never'
```

**Comment by charliermarsh** [comment]
(This works because with overrides, we just replace _all_ requirements of `typing-extensions` with whatever is in the overrides file.)

**Comment by leorochael** [comment]
> (This works because with overrides, we just replace _all_ requirements of `typing-extensions` with whatever is in the overrides file.)

I'm surprised this work, because I'd expect it would result in the resolution being unsolvable...

And I'm slightly suspicious of relying on it and it being just accidental behaviour that might be "fixed" in the future...

**Comment by charliermarsh** [comment]
@leorochael -- Can you say more about why you would expect that to be unsolvable? I would actually consider it a bug if the resolver failed there, rather than the other way around.

**Comment by leorochael** [comment]
I might be reading it wrong, but for me an overrides file containing:

* `typing-extensions ; sys_platform == 'never'`

Mean:

* Only consider this `typing-extensions` version: the one that cannot be installed on your platform, because your platform is not `never`.

And if there are packages depending on `typing-extensions`, in my mind that would mean their dependencies would be unsolvable because of that.

To me, declaring a constraints override on an uninstallable package is different that pretending a package is already installed.

**Comment by leorochael** [comment]
Ah, but I see where my logic is failing.

The constraints override doesn't say: "depend on this uninstallable package", it's saying, when depending on `typing-extensions`, only do that if the platform is `never`, which never happens.

So, yeah, as long as we add a test for that and document it, then yes, that would already be supported.

**Comment by hauntsaninja** [comment]
Oh, neat trick!

**Comment by Avasam** [comment]
> I realized that this actually is possible today... You can use a never-truthy marker. For example, to remove `typing-extensions`, use an `overrides.txt` like:
> 
> ```
> typing-extensions ; sys_platform == 'never'
> ```

Neat! However it's giving warnings in recent uv versions:
`warning: Missing version constraint (e.g., a lower bound) for typing-extensions`

I'd expect a "dependency elimination" to not warn about that.

**Comment by zanieb** [comment]
Yeah that seems incorrect. I wouldn't expect overrides to throw that warning at all (even if they're not an "elimination") cc @konstin

**Comment by tmm1** [comment]
Is there a way to override the version of a transitive dependency, instead of removing it?

For example, if I have a venv with pytorch-nightly installed, then want to add `vllm`, it fails because that project has a `torch == 2.5.1` requirement listed.

```
❯ uv add vllm==0.7.1 
  × No solution found when resolving dependencies:
  ╰─▶ Because vllm==0.7.1 depends on torchvision==0.20.1 and testproject depends on torchvision==0.22.0.dev20250124+cu126, we can conclude that testproject and vllm==0.7.1 are incompatible.
      And because testproject depends on vllm==0.7.1 and your project requires testproject, we can conclude that your project's requirements are unsatisfiable.
```

**Comment by notatallshaw** [comment]
> Is there a way to override the version of a transitive dependency, instead of removing it?

That's the whole purpose over dependency overrides:

https://docs.astral.sh/uv/reference/settings/#override-dependencies
https://docs.astral.sh/uv/pip/compile/#overriding-dependency-versions
https://docs.astral.sh/uv/concepts/resolution/#dependency-overrides

**Comment by hauntsaninja** [comment]
Btw, these days I actually find using uv's `dependency-metadata` to lie about the dependencies of my dependencies works better for me than overrides: https://docs.astral.sh/uv/reference/settings/#dependency-metadata

This is because it's narrower (i.e. override edges vs nodes as discussed in https://github.com/astral-sh/uv/issues/4422#issuecomment-2249044193 ). It's another very nice uv trick that flies under the radar a little :-)

**Comment by t-kalinowski** [comment]
Is there a way to override a dependency with `uv run`? Setting `UV_OVERRIDE` or `UV_CONSTRAINT` doesn’t seem to work for me.

Here’s the issue I’m running into:

I’m using both PyTorch and TensorFlow (with tensorflow-text). I want PyTorch to be GPU-capable and TensorFlow to run on CPU only. To do that, I explicitly declare `tensorflow-cpu`. However, `tensorflow-text` pulls in `tensorflow`, so I end up with both `tensorflow` and `tensorflow-cpu` being installed.

When running with `uv run`, this leads to a race condition where it's unpredictable which version of TensorFlow gets used.

Ideally, I’d like to tell `uv`: please don’t install `tensorflow`; treat `tensorflow-cpu` as satisfying the requirement from `tensorflow-text`.

This solution would ideally be accessible via `uv run` command line arguments, because that’s the mechanism that [reticulate](https://posit.co/blog/reticulate-1-41/) uses to resolve ephemeral venvs.

E.g., something like
```bash
uv run --no-project --no-cache --python 3.11.11 \
  --with tensorflow-cpu \
  --with tensorflow-text \
  --with 'tensorflow; from=tensorflow-cpu' \
-- python -c 'import tensorflow as tf; tf.config.list_physical_devices()' 
```

**Comment by carolynsoo** [comment]
> A use case where you need this functionality is when a package depends on opencv.

So this is exactly my current issue with uv.. I want only `opencv-python-headless` (on some machines that don't have `libGL.so`), but I depend on wheels that include `opencv-python`.

In root pyproject.toml I've set `sys_platform == 'never'"` for `opencv-python` in `override-dependencies`, but there are still files in various packages (each package has its own `pyproject.toml`, where I tried adding `opencv-python-headless`) that still complain `ModuleNotFoundError: No module named 'cv2'`. Am I missing a step?

**Comment by collinanderson** [comment]
I'm finding that even with the `sys_platform == 'never'` trick, uv still downloads's the package metadata, and the package still ends up with a version number the `uv.lock` file, and therefore would still get dependabot notifications for the dependency. (In my case `nodeenv` when installing `pyright[nodejs]`) Would be nice to have some sort of dependency exclusion list.

**Comment by charliermarsh** [comment]
Please provide a complete reproduction — hard to help without more details.

**Comment by collinanderson** [comment]
Steps to reproduce:
```
uv init tmp
cd tmp
echo -e "[tool.uv]\n override-dependencies = [ \"nodeenv; sys_platform == 'never'\" ]" >> pyproject.toml
uv add 'pyright[nodejs]'
grep -1 nodeenv uv.lock
```
Includes this in `uv.lock` which I'd like to avoid if possible:
```
[[package]]
name = "nodeenv"
version = "1.9.1"
source = { registry = "https://pypi.org/simple" }
sdist = { url = "https://files.pythonhosted.org/packages/43/16/fc88b08840de0e0a72a2f9d8c6bae36be573e475a6326ae854bcc549fc45/nodeenv-1.9.1.tar.gz", hash = "sha256:6ec12890a2dab7946721edbfbcd91f3319c6ccc9aec47be7c7e6b7011ee6645f", size = 47437, upload-time = "2024-06-04T18:44:11.171Z" }
wheels = [
    { url = "https://files.pythonhosted.org/packages/d2/1d/1b658dbd2b9fa9c4c9f32accbfc0205d532c8c6194dc0f2a4c0428e7128a/nodeenv-1.9.1-py2.py3-none-any.whl", hash = "sha256:ba11c9782d29c27c70ffbdda2d7415098754709be8a7056d79a737cd901155c9", size = 22314, upload-time = "2024-06-04T18:44:08.352Z" },
]
```

**Comment by ncoghlan** [comment]
The "use overrides to prevent installation" use case came up again recently in https://github.com/astral-sh/uv/issues/2500#issuecomment-3200409272 (as a way to instruct `uv` to skip installing packages available elsewhere on `sys.path` into the environment currently being updated)

**Comment by LTMeyer** [comment]
> Btw, these days I actually find using uv's `dependency-metadata` to lie about the dependencies of my dependencies works better for me than overrides in most cases: https://docs.astral.sh/uv/reference/settings/#dependency-metadata
> 
> This is because it's narrower (i.e. override edges vs nodes as discussed in [#4422 (comment)](https://github.com/astral-sh/uv/issues/4422#issuecomment-2249044193) ). It's another very nice uv trick that flies under the radar a little :-)

Would you have an example @hauntsaninja?

I was following [dependency-metadata documentation](https://docs.astral.sh/uv/reference/settings/#dependency-metadata). I added the following to my `pyproject.toml` assuming it will override dependencies for `<a-dependency-package>` (whereas `override-dependencies` overrides for all dependency packages as far as I understood).
```toml
[tool.uv]
dependency-metadata = [
    { name = "<a-dependency-package>", requires-dist = ["<b-package>>=2"]},
]
```
However, when installing through `uv sync --all-extras --all-groups --dev` I get an older `<b-package>` version installed, the one that is listed as dependency for the dependency package. I would expect to have `<b-package>` version 2 or more being installed. What did I miss?

Note that the dependency on `<a-dependency-package>` is defined as a development dependency (within a group https://docs.astral.sh/uv/concepts/projects/dependencies/#dependency-groups) instead of an optional dependency.

**Comment by konstin** [comment]
A reference can be PDM (https://pdm-project.org/latest/usage/config/#exclude-specific-packages-and-their-dependencies-from-the-lock-file):

> Sometimes you don't even want to include certain packages in the 
locked file because you are sure they won't be used by any code. In this
 case, you can completely skip them and their dependencies during 
dependency resolution:
> ```toml
> [tool.pdm.resolution]
> excludes = ["requests"]
> ```
> 
> With this config, <code>requests</code> will not be locked in the lockfile, and its dependencies such as <code>urllib3</code> and <code>idna</code>
 will also not show up in the resolution result, if not depended on by 
other packages. The installer will not be able to pick them up either.

**Comment by charliermarsh** [comment]
We just added support for excluding dependencies entirely by name: https://github.com/astral-sh/uv/pull/16528 (available in the next release)

**Comment by oddbookworm** [comment]
I would love a way for packages themselves to tell uv that they provide a complete replacement for a different package, reducing the need for users to put manual exclusions into their own pyproject.toml files. I'm one of the maintainers of [pygame-ce](https://github.com/pygame-community/pygame-ce), and our package is mutually exclusive with the original pygame (we install to the same location to keep the same import statement). We do provide nearly 100% backwards compatibility (the only difference that I can recall is a weird edge case that I don't know if anyone even uses), but users who want to use 3rd party libraries that specify `pygame` as a dependency have a fun task ahead of them when it comes to needing to uninstall pygame, then install pygame-ce.

**Comment by tamird** [comment]
Quoting @charliermarsh  from above:

> We just added support for excluding dependencies entirely by name

We tried replacing the never-truthy override with `exclude-dependencies`, but found a subtle difference in semantics: `exclude-dependencies` removes dependency edges, but does not eliminate a package supplied as a direct/root requirement.

With uv 0.11.17, given:

```toml
exclude-dependencies = ["chz"]
```

this still resolves the public package from a direct input:

```console
$ printf "chz\n" | uv pip compile - --config-file uv.toml --no-header
chz==0.4.0
typing-extensions==4.15.0
    # via chz
```

By contrast, the earlier never-truthy marker suggestion:

```toml
override-dependencies = ["chz; sys_platform == 'never'"]
```

emits no `chz` requirement for the same direct input, since the override replaces all requirements for that node.

Our use case is a fail-safe against ever selecting a public package with the same name as a first-party project, including accidental direct declarations, so `exclude-dependencies` cannot replace the so-called hack for us. Is excluding root requirements intentionally out of scope? If so, this distinction may be worth documenting; otherwise, would you consider making exclusions apply to direct requirements too?

**Comment by charliermarsh** [comment]
Will fix.
